### PR TITLE
Focus web action

### PR DESCRIPTION
**Pull request by xsmrcek**
- Sometimes we have to focus element, for example:
   - Before clicking on it
   - Before taking screenshot

### CODE REVIEW

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/189/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/FocusWebAction.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    @Override
+    public String name() {
+        return "FOCUS";
+    }
+
+    @Override
+    public void perform(WebActionContext ctx) {
+        WebElement element = findElement(ctx);
+
+        scrollIntoView(ctx, element);
```
Comment (thradec) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/FocusWebAction.java]: AFAIK "scroll into view" is not the same as "focus", so either is this action badly named or implemented

Comment (thradec) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/FocusWebAction.java]: is it possible to add test for it ?

Comment (paveljandejsek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/FocusWebAction.java]: If I remember correctly scrollIntoView uses JavaScript - are we able to do implementation without it to support more webdrivers?

Comment (xsmrcek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/FocusWebAction.java]: I rewrited action to use Selenium and renamed it.
But I wasn't able to test it, so to test it I added web property to verify is element is displayed on screen. And it is using javascript again :(

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/189/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/189/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/189/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/ScrollOnWebAction.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.interactions.Actions;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ScrollOnWebAction extends AbstractWebAction {
+
+    @Override
+    public String name() {
+        return "SCROLL_ON";
```
Comment (thradec) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/ScrollOnWebAction.java]: I'd keep original selenium wording, `MOVE_TO`, to simplified understanding what it really do.

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/property/DefaultWebProperties.java [hunk 3]
```diff
@@ -83,4 +89,41 @@ public Boolean value(WebPropertyContext ctx) {
...
+                                "  cx = box.left + box.width / 2,         " +
+                                "  cy = box.top + box.height / 2,         " +
+                                "  e = document.elementFromPoint(cx, cy); " +
+                                "for (; e; e = e.parentElement) {         " +
+                                "  if (e === elem)                        " +
+                                "    return true;                         " +
+                                "}                                        " +
+                                "return false;                            "
+                            , element);
+                        return (isVisible && visibleOnScreen);
```
Comment (thradec) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/property/DefaultWebProperties.java]: minor: I'd use some form of "display" in the naming of local variables when the whole property/class use the name DISPLAY

**Comment by xsmrcek** [comment]
I did try implemetation with Selenium moveToElement() method and ended with this exception: MoveTargetOutOfBoundsException. 
So I am back at java script implementation.

I resolved all old conversations as I tryed to implement all comments.

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/ScrollOnWebAction.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    public String name() {
+        return "MOVE_TO";
+    }
+
+    @Override
+    public void perform(WebActionContext ctx) {
+        WebElement element = findElement(ctx);
+        scrollIntoView(element);
+    }
+}
```
Comment (thradec) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/web/action/ScrollOnWebAction.java]: I'd prefer to use only one term "scroll on" or "move to"

File: jbehave-support-core/src/test/groovy/org/jbehavesupport/core/BrowserStackIT.groovy [hunk 5]
```diff
@@ -100,5 +102,6 @@ class BrowserStackIT extends Specification implements TestSupportBrowserStack {
         "sample/WebWaitCondition.story"         | _
         //"sample/WebAction.story"                | _
         "sample/WebNavigation.story"            | _
+        "sample/WebFocusAction.story"           | _
```
Comment (xsmrcek) [inline jbehave-support-core/src/test/groovy/org/jbehavesupport/core/BrowserStackIT.groovy]: Wrong story...

**Pull request merged**